packages/gatsby-source-contentful/src/on-pre-bootstrap.js

```diff
@@ -0,0 +1,356 @@
+    {
+      parentSpan,
+    }
+  )
+  processingActivity.start()
+
+  // Create a map of up to date entries and assets
+  function mergeSyncData(previous, current, deleted) {
+    const entryMap = new Map()
+    previous.forEach(e => !deleted.has(e.sys.id) && entryMap.set(e.sys.id, e))
+    current.forEach(e => !deleted.has(e.sys.id) && entryMap.set(e.sys.id, e))
+    return [...entryMap.values()]
+  }
+
+  const mergedSyncData = {
+    entries: mergeSyncData(
+      previousSyncData.entries,
+      currentSyncData.entries,
+      new Set(currentSyncData.deletedEntries.map(e => e.sys.id))
+    ),
+    assets: mergeSyncData(
+      previousSyncData.assets,
+      currentSyncData.assets,
+      new Set(currentSyncData.deletedAssets.map(e => e.sys.id))
+    ),
+    tagItems,
+  }
+
+  // Store a raw and unresolved copy of the data for caching
+  const mergedSyncDataRaw = _.cloneDeep(mergedSyncData)
+
+  // Use the JS-SDK to resolve the entries and assets
+  const res = createClient({
+    space: `none`,
+    accessToken: `fake-access-token`,
+  }).parseEntries({
+    items: mergedSyncData.entries,
+    includes: {
+      assets: mergedSyncData.assets,
+      entries: mergedSyncData.entries,
+    },
+  })
+
+  mergedSyncData.entries = res.items
+
+  // Inject raw API output to rich text fields
+  const richTextFieldMap = new Map()
+  contentTypeItems.forEach(contentType => {
+    richTextFieldMap.set(
+      contentType.sys.id,
+      contentType.fields
+        .filter(field => field.type === `RichText`)
+        .map(field => field.id)
+    )
+  })
+
+  const rawEntries = new Map()
+  mergedSyncDataRaw.entries.forEach(rawEntry =>
+    rawEntries.set(rawEntry.sys.id, rawEntry)
+  )
+
+  mergedSyncData.entries.forEach(entry => {
+    const contentTypeId = entry.sys.contentType.sys.id
+    const richTextFieldIds = richTextFieldMap.get(contentTypeId)
+    if (richTextFieldIds) {
+      richTextFieldIds.forEach(richTextFieldId => {
+        if (!entry.fields[richTextFieldId]) {
+          return
+        }
+        entry.fields[richTextFieldId] = rawEntries.get(entry.sys.id).fields[
+          richTextFieldId
+        ]
+      })
+    }
+  })
+
+  // @todo based on the sys metadata we should be able to differentiate new and updated entities
+  reporter.info(
+    `Contentful: ${currentSyncData.entries.length} new/updated entries`
+  )
+  reporter.info(
+    `Contentful: ${currentSyncData.deletedEntries.length} deleted entries`
+  )
+  reporter.info(`Contentful: ${previousSyncData.entries.length} cached entries`)
+  reporter.info(
+    `Contentful: ${currentSyncData.assets.length} new/updated assets`
+  )
+  reporter.info(`Contentful: ${previousSyncData.assets.length} cached assets`)
+  reporter.info(
+    `Contentful: ${currentSyncData.deletedAssets.length} deleted assets`
+  )
+
+  // Remove deleted entries & assets
+  reporter.verbose(`Removing deleted Contentful entries & assets`)
+
+  // @todo this should happen when sourcing?
```

Comment:
Is deleting of nodes on pre bootstrap a bad idea? We might move it to sourceNodes 🤷

Reply:
yes :) you have to do all node creation/deletions in source-nodes